### ISSUE TITLE
`fetch` POST request breaks for resource route with loader + default component

### ISSUE THREAD

**Issue by jmaldon1** [remix-run/remix#6941]
### What version of Remix are you using?

1.19.1

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

```tsx
//_resources.my-resource-route.tsx

export const loader = async (_: LoaderArgs): Promise<never> => {
  throw new Response("Not Found", {
    status: 404,
  });
};

export const action = async ({
  request,
}: ActionArgs): Promise<TypedResponse<string>> => {
  return json("HELLO");
};

export default function Component(): null {
  return null;
}
```

I write my resources in this manner so that if a user managed to visit this page accidently. It would show a 404 error message, like any other page that doesn't exist, like so:

![image](https://github.com/remix-run/remix/assets/25569545/069d0caa-275f-4fb0-b758-c94b3b8df08e)

However when you make a `fetch` POST request to this resource route 

```ts
const data = new FormData();
  data.append("type", "test");
  const response = await fetch("/my-resource-route", {
    method: "POST",
    headers: {
      accept: "application/json",
    },
    body: data,
  });
  console.log(await response.text());
```

you will just get the following HTML response:

```html
<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width,initial-scale=1"/><title>404 </title><link rel="manifest" href="/resources/manifest.webmanifest"/><link rel="preload stylesheet" href="/build/_assets/nprogress-WBGJMCZ7.css" as="style"/><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&amp;display=swap"/><meta name="emotion-insertion-point" content="emotion-insertion-point"/><style data-emotion="css-global 0"></style><style data-emotion="css-global o6gwfi">html{-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;box-sizing:border-box;-webkit-text-size-adjust:100%;}*,*::before,*::after{box-sizing:inherit;}strong,b{font-weight:700;}body{margin:0;color:rgba(0, 0, 0, 0.87);font-family:"Roboto","Helvetica","Arial",sans-serif;font-weight:400;font-size:1rem;line-height:1.5;letter-spacing:0.00938em;background-color:#fff;}@media print{body{background-color:#fff;}}body::backdrop{background-color:#fff;}</style><style data-emotion="css 17nso2j-MuiContainer-root">.css-17nso2j-MuiContainer-root{width:100%;margin-left:auto;box-sizing:border-box;margin-right:auto;display:block;min-height:100vh;border:1px solid;border-color:rgba(0, 0, 0, 0.12);}@media (min-width:1200px){.css-17nso2j-MuiContainer-root{max-width:1200px;}}@media (min-width:0px){.css-17nso2j-MuiContainer-root{max-width:600px;}}@media (min-width:900px){.css-17nso2j-MuiContainer-root{max-width:900px;}}</style></head><body><div class="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-17nso2j-MuiContainer-root"><div class="MuiBox-root css-0"><h1>404<!-- -->: </h1><p>Oops! Looks like you tried to visit a page that does not exist.</p></div></div><script>((STORAGE_KEY, restoreKey) => {
    if (!window.history.state || !window.history.state.key) {
      let key = Math.random().toString(32).slice(2);
      window.history.replaceState({
        key
      }, "");
    }
    try {
      let positions = JSON.parse(sessionStorage.getItem(STORAGE_KEY) || "{}");
      let storedY = positions[restoreKey || window.history.state.key];
      if (typeof storedY === "number") {
        window.scrollTo(0, storedY);
      }
    } catch (error) {
      console.error(error);
      sessionStorage.removeItem(STORAGE_KEY);
    }
 ......
```

Basically you get the 404 HTML response page that I show to users. Almost as if I was making a GET request to the `my-resource-route`

If I comment out component this issue goes away and the fetch request works:

```tsx
// export default function Component(): null {
//   return null;
// }
```

except that now when a user visits the page via a browser, they aren't shown the styled 404 response, they just see this:

![image](https://github.com/remix-run/remix/assets/25569545/4860decc-1b72-4819-a13b-6832a8b3a218)

### Expected Behavior

`Fetch` POST requests should work normally when writing resource routes in this manner.

### Actual Behavior

Shown in the steps to reproduce

**Comment by brophdawg11** [comment]
This is working as expected since resource routes are not expected to export a default export.  If a route module has a default exported component, then it is not a resource route.  The presence of the default export is what tells Remix that this is a UI route that should return HTML.  If you need to hit this as an API _and_ you need to export a component for 404 handling, then I would suggest hitting it with a `fetcher` will which will correctly add the `_data` param to enforce the returning of JSON.

**Comment by jmaldon1** [comment]
@brophdawg11 Makes sense. 

The reason I am using a `fetch` call is because I am doing it from a from a service worker. So I don't think the `fetcher` call is feasible there.
